### PR TITLE
PR 3 of 4 for SNV Consensus: Consensus file creator

### PR DESCRIPTION
**Pull request by cansavvy**
### Purpose/implementation

This is an analysis that originated from [this original PR](https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/174) but after some reorganization has been split up and spread out. 

This third PR takes the merged VAF and TMB files and creates the consensus file. This consensus file is a zipped TSV file which is "MAF-like" see more details in the README. 

**The input files:**
- `all_callers_vaf.rds` - contains all the VAF file information for all callers.
- `all_callers_tmb.rds` - contains all the TMB file information for all callers.
- `callers_per_mutation.rds` - contains a breakdown for each mutation of what callers called it. Will be used to identify the consensus mutations.

### Issue

For SNV caller comparison #161, #103 and Tumor Mutation Burden #3 and sort of #11

### Directions for reviewers

This script has to load in 3 of the files that were created in 03-merge_callers. This takes some time. Do you have suggestions for decreasing the run time for this script?
How do you feel about the options and error handling for this script? 

### Results

This script gives back the consensus mutation file, which at this point in time, is any mutations that have been called by Lancet, Mutect2, and Strelka2. 

#### Docker and continuous integration

These items were taken care of in #183
- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile. 
- [x] This analysis has been added to continuous integration.

#### PR Checklist
* [x] Run a linter
* [x] Set the seed- I didn't, but let me know if I should.
* [x] Comments and/or documentation up to date 
* [x] Double check your paths
* [x] Spell check any Rmd file or md file
* [x] Restart R and run all notebooks fresh and save

### CODE REVIEW

File: analyses/snv-callers/scripts/04-create_consensus_mut_files.R [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+    "A consensus TMB file already exists: \n",
+    consensus_tmb_file, "\n",
+    "Use --overwrite if you want to overwrite it."
+  ))
+} else {
+  # Set up BED region files for TMB calculations
+  wgs_bed <- readr::read_tsv(file.path(opt$bed_wgs), col_names = FALSE)
+  wxs_bed <- readr::read_tsv(file.path(opt$bed_wxs), col_names = FALSE)
+
+  # Calculate size of genome surveyed
```
Comment (jaclyn-taroni) [inline analyses/snv-callers/scripts/04-create_consensus_mut_files.R]: Can you put a detailed comment in here about the indices please? I see that there are not column names.

File: analyses/snv-callers/scripts/04-create_consensus_mut_files.R [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+`%>%` <- dplyr::`%>%`
+
+# Load library:
+library(optparse)
+
+################################ Set up options ################################
+# Set up optparse options
+option_list <- list(
+  make_option(
+    opt_str = c("-m", "--merged_files"), type = "character",
```
[the remark this answers is not in the document]
Comment (cansavvy) [reply analyses/snv-callers/scripts/04-create_consensus_mut_files.R]: Yes. That's a more accurate name.

File: analyses/snv-callers/scripts/04-create_consensus_mut_files.R [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+)
+
+needed_files <- c(
+  file.path(root_dir, opt$merged_files, paste0(needed_files, opt$file_format)),
+  opt$bed_wgs,
+  opt$bed_wxs
+)
+
+# Get list of which files were found
+files_found <- sapply(needed_files, file.exists)
```
[the remark this answers is not in the document]
Comment (cansavvy) [reply analyses/snv-callers/scripts/04-create_consensus_mut_files.R]: This logic was something I used in 02-run_eval, but it is not working as it is supposed to. I will file another PR to fix some of these things in 02-run_eval.R

**Comment by cansavvy** [comment]
@jaclyn-taroni Ready for another gander.

File: analyses/snv-callers/scripts/04-create_consensus_mut_files.R [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+  if (any(existing_files)) {
+    warning(cat(
+      "Some output files already exist and will not be overwritten unless you use --overwrite: \n",
+      paste0(output_files[which(existing_files)], "\n")
+    ))
+  }
+}
+############################### Read in the files ##############################
+# Read in merged VAF_df
+vaf_df <- readr::read_rds(needed_files[[1]])
```
Comment (jaclyn-taroni) [inline analyses/snv-callers/scripts/04-create_consensus_mut_files.R]: So the output of `03-merge_callers.R` will always be RDS, got it.

File: analyses/snv-callers/scripts/04-create_consensus_mut_files.R [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+if (file.exists(consensus_mut_zip_file) && !opt$overwrite) {
+  # Stop if this file exists and overwrite is set to FALSE
+  warning(cat(
+    "A zipped consensus mutation file already exists: \n",
+    consensus_mut_zip_file, "\n",
+    "Use --overwrite if you want to overwrite it."
+  ))
+} else {
+  # Zip this file up.
+  zip(consensus_mut_zip_file, consensus_mut_file)
```
Comment (jaclyn-taroni) [inline analyses/snv-callers/scripts/04-create_consensus_mut_files.R]: I'm not going to ask you to change this but I would have just stuck this in the logic above. If you're generating the new tsv file, you should generate the new zip file.

File: analyses/snv-callers/scripts/04-create_consensus_mut_files.R [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+
+  # Calculate size of genome surveyed
+  # These files are BED files where the third column is the End position and 
+  # the second column is the Start position. 
+  # So End - Start gives the size of each range. Sum the gives the total size in bp.
+  wgs_genome_size <- sum(wgs_bed[, 3] - wgs_bed[, 2])
+  wxs_exome_size <- sum(wxs_bed[, 3] - wxs_bed[, 2])
+
+  # Calculate TMBs and write to TMB file
+  tmb_df <- calculate_tmb(vaf_df,
```
Comment (jaclyn-taroni) [inline analyses/snv-callers/scripts/04-create_consensus_mut_files.R]: This should be `consen_mutation`

File: analyses/snv-callers/scripts/04-create_consensus_mut_files.R (hunk 4 above)
Comment (cansavvy) [reply analyses/snv-callers/scripts/04-create_consensus_mut_files.R]: Yes. I figure people don't need options at this point. It's mainly just an intermediary file anyway. At a later point I want to institute a change so the file doesn't have to be written and loaded again but for now this is fine.

File: analyses/snv-callers/scripts/04-create_consensus_mut_files.R (hunk 6 above)
Comment (cansavvy) [reply analyses/snv-callers/scripts/04-create_consensus_mut_files.R]: 🦅

**Comment by cansavvy** [comment]
Made those changes, what do we think, @jaclyn-taroni ?

**Pull request merged**